pallets/did/src/lib.rs

```diff
@@ -28,93 +28,364 @@ mod tests;
 use frame_system::{self, ensure_signed};
-use sp_runtime::{codec::Codec, traits::Member};
-use sp_std::prelude::*;
+use sp_runtime::traits::Verify;
+use sp_std::{prelude::{Clone}, convert::TryFrom, collections::{btree_set::BTreeSet}, vec::Vec};
+use sp_core::{ed25519, sr25519};
+
+/// Type of a payload of data (to verify signatures against).
+pub type Payload = Vec<u8>;
+
+/// Type of a signature.
+pub type Signature = Vec<u8>;
+
+/// Type for a raw key (either verification or encryption key).
+pub type KeyRawValue = Vec<u8>;
+
+/// Type for a DID identifier.
+pub type DIDIdentifier = Vec<u8>;
+
+/// Type for a public key identifier.
+pub type KeyIdentifier = Vec<u8>;
+
+/// Type for a URL.
+pub type URL = Vec<u8>;
+
+/// Type for the result of verifying a signature given a verification key.
+pub type SignatureVerificationResult = Result<bool, SignatureError>;
+
+/// Type for the result of verifying the signature of a DID subject for a given DID operation.
+pub type DIDSignatureVerificationResult = Result<bool, DIDError>;
 
-/// The DID trait
 pub trait Config: frame_system::Config {
-	/// DID specific event type
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+	/// DID specific event type.
+	type Event: From<Event> + Into<<Self as frame_system::Config>::Event>;
 
 	/// Weight information for extrinsics in this pallet.
 	type WeightInfo: WeightInfo;
+}
+
+/// Trait representing a public key under the control of a DID subject.
+pub trait DIDPublicKey {
+	/// Returns the key method description as in the [DID specification](https://w3c.github.io/did-spec-registries/#verification-method-types).
+	fn get_did_key_description(&self) -> &'static str;
+}
+
+/// Enum representing the types of verification keys a DID can control.
+#[derive(Clone, Decode, Debug, Encode, Eq, Ord, PartialEq, PartialOrd)]
+pub enum PublicVerificationKey {
+	/// An Ed25519 public key.
+	Ed25519([u8; 32]),
+	/// A Sr25519 public key.
+	Sr25519([u8; 32]),
+}
+
+impl PublicVerificationKey {
+	/**
+	Given a payload and a signature, the specific public verification key will return either
+	a SignatureVerificationResult if the signature is not properly formed, or a boolean indicating
+	the result of the verification.
+	 */
+	fn verify_signature(&self, payload: &Payload, signature: &Signature) -> SignatureVerificationResult {
+		// Discard all invalid signatures by comparing them with the expected length of the specific verification key.
+		ensure!(signature.len() == self.get_expected_signature_size(), SignatureError::InvalidSignatureFormat);
+		
+		// Signature::try_from expects a &[u8], so we cast the vec (of the right size) to the expected type.
+		let signature = &signature as &[u8];
+		// TODO: did not find a way to return a Signature object from the match and then call directly verify on that (requires to return a Trait object)...
+		match self {
```

Comment:
I had a hard time returning a Box<dyn PublicVerificationKey> from within the match...